torch/csrc/autograd/python_function.cpp

```diff
@@ -146,35 +159,71 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
   }
 
   // Massage the Python results tuple back into a C++ variable_list
-  variable_list results;
-  results.reserve(num_outputs);
-  for (int i = 0; i != num_outputs; ++i) {
-    PyObject* output = PyTuple_GET_ITEM(r.get(), i);
-    bool was_variable = is_variable_input[i];
-    if (!was_variable) {
-      if (output != Py_None) {
-        std::string msg("function ");
-        msg += name() + " returned a gradient different than None at position ";
-        msg += std::to_string(i + 1) +
-            ", but the corresponding forward input was not a Variable";
-        throw std::runtime_error(msg);
-      }
+  return to_variable_list(r.get(), is_variable_input);
+}
+
+auto PyNode::compiled_apply(
+    variable_list&& inputs,
+    std::optional<PyObject*> compiler) -> variable_list {
+  pybind11::gil_scoped_acquire gil;
+  at::OptionalDeviceGuard _device_guard;
+  THPFunction* py_fn = (THPFunction*)obj;
+
+  // Massage a C++ variable_list into a Python arguments tuple
+  THPObjectPtr pyInputs(to_py_args(inputs, &_device_guard));
+
+  const auto& is_variable_input = py_fn->is_variable_input;
+  const auto& input_infos = py_fn->input_info;
+  // input_info only contains info from variable inputs and should be a subset
+  TORCH_INTERNAL_ASSERT(is_variable_input.size() >= input_infos.size());
+
+  // The gradients returned in the backwards should match the number of inputs
+  // to the forward, and their shapes, so we pass the fwdInputs
+  THPObjectPtr fwdInputs(
+      PyTuple_New(static_cast<Py_ssize_t>(is_variable_input.size())));
+  int offset = 0;
+  for (const auto i : c10::irange(is_variable_input.size())) {
+    if (!is_variable_input[i]) {
+      // input at i is not a variable, add an offset
+      PyTuple_SET_ITEM(fwdInputs.get(), i, Py_None);
+      offset++;
       continue;
     }
-    if (output == Py_None) {
-      results.emplace_back();
-    } else {
-      if (!THPVariable_Check(output)) {
-        std::string msg("expected Variable or None (got ");
-        msg += THPUtils_typename(output);
-        msg += ")";
-        throw std::runtime_error(msg);
-      }
-      results.emplace_back(THPVariable_Unpack(output));
-    }
+
+    const auto& input_info = input_infos[i - offset];
+
+    // TODO: figure out how to pass symint sizes directly and remove this python
+    // call
+    auto zeros_without_gil = [](const VariableInfo& variable,
+                                at::OptionalDeviceGuard& dg) {
+      pybind11::gil_scoped_release gil;
+      return variable.zeros(dg);
+    };
+    PyTuple_SET_ITEM(
+        fwdInputs.get(),
+        static_cast<Py_ssize_t>(i),
+        THPVariable_Wrap(zeros_without_gil(input_info, _device_guard)));
```

Comment:
Allocating zero tensors here will add nodes to the graph that don't belong in the graph.  It also makes the `maybe_clone()` called on these in python a bit redundant.  

We should either:
1) Pass the sizes/metadata into `proxy_call_backward` and allocate these tensors in python; or
2) Disable tracing here so they aren't captured in the graph and remove the `maybe_clone` in python (since they are already fresh tensors).

Reply:
changed to pass as THPSize, no longer appears in the graph